routing.go

```diff
@@ -500,132 +421,154 @@ func (dht *IpfsDHT) FindProviders(ctx context.Context, c cid.Cid) ([]peer.AddrIn
 	return providers, nil
 }
 
-// FindProvidersAsync is the same thing as FindProviders, but returns a channel.
-// Peers will be returned on the channel as soon as they are found, even before
-// the search query completes. If count is zero then the query will run until it
-// completes. Note: not reading from the returned channel may block the query
-// from progressing.
-func (dht *IpfsDHT) FindProvidersAsync(ctx context.Context, key cid.Cid, count int) <-chan peer.AddrInfo {
+// FindProvidersAsyncExtended searches until the context expires.
+func (dht *IpfsDHT) FindProvidersAsyncExtended(ctx context.Context, key cid.Cid, opts ...routing.Option) (<-chan peer.AddrInfo, <-chan []peer.ID, error) {
 	if !dht.enableProviders || !key.Defined() {
-		peerOut := make(chan peer.AddrInfo)
+		peerOut, closestPeers := make(chan peer.AddrInfo), make(chan []peer.ID)
 		close(peerOut)
-		return peerOut
+		close(closestPeers)
+		return peerOut, closestPeers, routing.ErrNotSupported
 	}
 
-	chSize := count
-	if count == 0 {
-		chSize = 1
+	var cfg routing.Options
+	if err := cfg.Apply(opts...); err != nil {
+		return nil, nil, err
 	}
-	peerOut := make(chan peer.AddrInfo, chSize)
 
 	keyMH := key.Hash()
 
-	go dht.findProvidersAsyncRoutine(ctx, keyMH, count, peerOut)
-	return peerOut
+	processors := dhtrouting.GetProcessors(&cfg)
+	if processors == nil {
+		newValuesOnly := &dhtrouting.NewPeerIDFilter{
+			Key:   string(keyMH),
+			Peers: peer.NewSet(),
+		}
+
+		quorum := &dhtrouting.CountStopper{
+			MaxCount: dhtrouting.GetQuorum(&cfg),
+		}
+
+		processors = []dhtrouting.Processor{newValuesOnly, quorum}
```

Comment:
Note: While SearchValue does quorum -> bestValue, here we do newValuesOnly -> quorum. I'd like some comments explaining the pipeline order for each function.